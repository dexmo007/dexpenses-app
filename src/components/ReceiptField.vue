<template>
  <div>
    <v-icon :color="missing && required ? 'red lighten-1' : undefined">{{icon}}</v-icon>
    <span class="receipt-field-value">
      <span
        class="fixed"
        v-dblclick-or-hold
        @dblclickorhold="edit"
        v-show="!editing"
      >{{displayValue}}</span>
      <v-form
        class="form"
        ref="form"
        @submit.prevent="submit"
        v-show="editing || missing"
        v-model="valid"
      >
        <slot
          name="editor"
          :value="changedValueHolder"
          :loading="submitted"
        >
          <v-text-field
            class="edit-input receipt-field-text-field"
            height="1em"
            ref="editInput"
            v-model="changedValueHolder.changedValue"
            :loading="submitted"
            :rules="rules"
            :type="type"
            :mask="mask"
            :return-masked-value="true"
          />
        </slot>
        <v-btn
          flat
          icon
          type="submit"
          v-if="editing || changedValueHolder.changedValue"
        >
          <v-icon>check</v-icon>
        </v-btn>
        <v-btn
          flat
          icon
          v-if="editing || changedValueHolder.changedValue"
          @click="editing = false; changedValueHolder.changedValue = value"
        >
          <v-icon>close</v-icon>
        </v-btn>
      </v-form>
    </span>
  </div>
</template>
<script>
import { mapActions } from 'vuex';
import deepEqual from 'deep-equal';

export default {
  name: 'ReceiptField',
  props: {
    icon: String,
    receipt: Object,
    field: String,
    filter: String,
    required: Boolean,
    rules: Array,
    type: String,
    mask: String,
    parser: {
      type: Function,
      default: v => (v ? v.trim() : v),
    },
    formatter: {
      type: [Function, String],
      default: v => (!v ? '' : v.toString()),
    },
  },
  data() {
    return {
      changedValueHolder: {
        changedValue: null,
      },
      submitted: false,
      editing: false,
      valid: false,
    };
  },
  computed: {
    value() {
      return this.receipt.result.data[this.field];
    },
    displayValue() {
      if (this.filter) {
        return this.$options.filters[this.filter](this.value);
      }
      return this.value;
    },
    missing() {
      return !this.value;
    },
  },
  methods: {
    ...mapActions('receipts', ['updateReceipt']),
    edit() {
      this.editing = true;
      this.changedValueHolder.changedValue = this.format(this.value);
      this.$nextTick(
        () => this.$refs.editInput && this.$refs.editInput.focus()
      );
    },
    async submit() {
      if (!this.valid) {
        return;
      }
      this.editing = false;
      let value;
      if (
        !this.changedValueHolder.changedValue ||
        !this.changedValueHolder.changedValue.trim()
      ) {
        value = null;
      } else {
        value = this.parser(this.changedValueHolder.changedValue);
      }
      this.changedValueHolder.changedValue = '';
      if (deepEqual(value, this.value)) {
        return;
      }
      this.submitted = true;
      await this.updateReceipt({
        id: this.receipt.id,
        field: this.field,
        value,
      });
      this.submitted = false;
    },
    format(v) {
      if (typeof this.formatter === 'string') {
        return this.$options.filters[this.formatter](v);
      }
      return this.formatter(v);
    },
  },
};
</script>
<style scoped>
.receipt-field-value {
  margin-left: 0.2em;
  display: flex;
  justify-content: center;
  align-items: center;
}
.form {
  display: flex;
  height: 1em;
}
</style>
<style>
.receipt-field-text-field {
  height: 1em;
  margin: 0;
  padding: 0;
}
.receipt-field-text-field input {
  width: 100%;
}
</style>
