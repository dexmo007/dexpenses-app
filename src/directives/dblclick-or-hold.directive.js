import Vue from 'vue';

Vue.directive('dblclickOrHold', {
  bind(el, binding) {
    /*
     * dblclick event
     */
    el.addEventListener('dblclick', () => {
      el.dispatchEvent(new Event('dblclickorhold'));
    });
    /*
     * Hold event
     */
    const holdDuration = binding.value || 500;
    let timeout;
    const holdStart = () => {
      timeout = setTimeout(() => {
        timeout = null;
        el.dispatchEvent(new Event('dblclickorhold'));
      }, holdDuration);
    };
    el.addEventListener('mousedown', holdStart);
    el.addEventListener('touchstart', e => {
      e.preventDefault();
      holdStart();
    });
    const cancel = () => {
      if (timeout) {
        clearTimeout(timeout);
        timeout = null;
      }
    };
    el.addEventListener('mouseup', cancel);
    el.addEventListener('mouseleave', cancel);
    el.addEventListener('touchend', cancel);
  },
});
