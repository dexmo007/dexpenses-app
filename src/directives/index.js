import './dblclick-or-hold.directive';