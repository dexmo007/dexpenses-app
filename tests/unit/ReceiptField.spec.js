import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
import createVuetify from './vuetify';
import ReceiptField from '@/components/receipts/ReceiptField.vue';
import '@/directives';

const localVue = createLocalVue();

describe('ReceiptField.vue', () => {
  let vuetify;
  beforeEach(() => {
    vuetify = createVuetify();
  });

  it('should render correctly', () => {
    const wrapper = shallowMount(ReceiptField, {
      localVue,
      vuetify,
      propsData: {
        field: 'field',
        icon: 'icon',
        receipt: {
          result: {
            data: {
              field: 'value',
            },
          },
        },
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should apply the provided filter', () => {
    const wrapper = shallowMount(ReceiptField, {
      localVue,
      vuetify,
      propsData: {
        field: 'field',
        icon: 'icon',
        receipt: {
          result: {
            data: {
              field: 'value',
            },
          },
        },
        filter: 'uppercase',
      },
      filters: {
        uppercase(v) {
          return v ? v.toUpperCase() : v;
        },
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should show red icon if required value is missing', () => {
    const wrapper = shallowMount(ReceiptField, {
      localVue,
      vuetify,
      propsData: {
        field: 'field',
        icon: 'icon',
        receipt: {
          result: {
            data: {
              field: undefined,
            },
          },
        },
        required: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should enter edit mode on double-click of the text', () => {
    const wrapper = mount(ReceiptField, {
      localVue,
      vuetify,
      propsData: {
        field: 'field',
        icon: 'icon',
        receipt: {
          result: {
            data: {
              field: 'value',
            },
          },
        },
      },
      stubs: {
        'v-form': true, // real mount of v-form causes inexplicable internal error
      },
    });
    expect(wrapper.element).toMatchSnapshot();
    wrapper.find('.receipt-field-value .fixed').trigger('dblclick');
    expect(wrapper.element).toMatchSnapshot();
  });
});
