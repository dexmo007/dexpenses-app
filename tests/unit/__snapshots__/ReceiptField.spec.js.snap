// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReceiptField.vue should apply the provided filter 1`] = `
<div>
  <v-icon-stub>
    icon
  </v-icon-stub>
   
  <span
    class="receipt-field-value"
  >
    <span
      class="fixed"
    >
      VALUE
    </span>
     
    <v-form-stub
      class="form"
      style="display: none;"
    >
      <v-text-field-stub
        backgroundcolor=""
        class="edit-input receipt-field-text-field"
        clearicon="$vuetify.icons.clear"
        errorcount="1"
        errormessages=""
        height="1em"
        loaderheight="2"
        messages=""
        rules=""
        successmessages=""
        type="text"
      />
       
      <!---->
       
      <!---->
    </v-form-stub>
  </span>
</div>
`;

exports[`ReceiptField.vue should enter edit mode on double-click of the text 1`] = `
<div>
  <i
    aria-hidden="true"
    class="v-icon notranslate material-icons theme--light"
  >
    icon
  </i>
   
  <span
    class="receipt-field-value"
  >
    <span
      class="fixed"
    >
      value
    </span>
     
    <v-form-stub
      class="form"
      style="display: none;"
    >
      <div
        class="v-input edit-input receipt-field-text-field theme--light v-text-field"
      >
        <div
          class="v-input__control"
        >
          <div
            class="v-input__slot"
            style="height: 1em;"
          >
            <div
              class="v-text-field__slot"
            >
              <input
                id="input-20"
                type="text"
              />
            </div>
          </div>
          <div
            class="v-text-field__details"
          >
            <div
              class="v-messages theme--light"
            >
              <span
                class="v-messages__wrapper"
                name="message-transition"
                tag="div"
              />
            </div>
          </div>
        </div>
      </div>
       
      <!---->
       
      <!---->
    </v-form-stub>
  </span>
</div>
`;

exports[`ReceiptField.vue should enter edit mode on double-click of the text 2`] = `
<div>
  <i
    aria-hidden="true"
    class="v-icon notranslate material-icons theme--light"
  >
    icon
  </i>
   
  <span
    class="receipt-field-value"
  >
    <span
      class="fixed"
      style="display: none;"
    >
      value
    </span>
     
    <v-form-stub
      class="form"
      style=""
    >
      <div
        class="v-input edit-input receipt-field-text-field v-input--is-label-active v-input--is-dirty theme--light v-text-field"
      >
        <div
          class="v-input__control"
        >
          <div
            class="v-input__slot"
            style="height: 1em;"
          >
            <div
              class="v-text-field__slot"
            >
              <input
                id="input-20"
                type="text"
              />
            </div>
          </div>
          <div
            class="v-text-field__details"
          >
            <div
              class="v-messages theme--light"
            >
              <span
                class="v-messages__wrapper"
                name="message-transition"
                tag="div"
              />
            </div>
          </div>
        </div>
      </div>
       
      <button
        class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
        type="submit"
      >
        <span
          class="v-btn__content"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate material-icons theme--light"
          >
            check
          </i>
        </span>
      </button>
       
      <button
        class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate material-icons theme--light"
          >
            close
          </i>
        </span>
      </button>
    </v-form-stub>
  </span>
</div>
`;

exports[`ReceiptField.vue should render correctly 1`] = `
<div>
  <v-icon-stub>
    icon
  </v-icon-stub>
   
  <span
    class="receipt-field-value"
  >
    <span
      class="fixed"
    >
      value
    </span>
     
    <v-form-stub
      class="form"
      style="display: none;"
    >
      <v-text-field-stub
        backgroundcolor=""
        class="edit-input receipt-field-text-field"
        clearicon="$vuetify.icons.clear"
        errorcount="1"
        errormessages=""
        height="1em"
        loaderheight="2"
        messages=""
        rules=""
        successmessages=""
        type="text"
      />
       
      <!---->
       
      <!---->
    </v-form-stub>
  </span>
</div>
`;

exports[`ReceiptField.vue should show red icon if required value is missing 1`] = `
<div>
  <v-icon-stub
    color="red lighten-1"
  >
    icon
  </v-icon-stub>
   
  <span
    class="receipt-field-value"
  >
    <span
      class="fixed"
    >
      
    </span>
     
    <v-form-stub
      class="form"
    >
      <v-text-field-stub
        backgroundcolor=""
        class="edit-input receipt-field-text-field"
        clearicon="$vuetify.icons.clear"
        errorcount="1"
        errormessages=""
        height="1em"
        loaderheight="2"
        messages=""
        rules=""
        successmessages=""
        type="text"
      />
       
      <!---->
       
      <!---->
    </v-form-stub>
  </span>
</div>
`;
